### ISSUE TITLE
Performance decrease by synchronous callback handling

### ISSUE THREAD

**Issue by long131ss** [continental/ecal#468]
**Describe the bug**

I am trying to integrate the ecal framework into cartographer_ros project now. In the cartographer node , subscribe the sensor message /scan (the frequency is 10hz),/odom (the frequency is 100 hz),/imu/data_raw(the frequency is 100 hz),/generic_tf/center_of_rotation_2_odom (the frequency is 100hz). 

I add all related events for all the subscribers in my node. 
```
    ::eCAL::ros::CSubscriber<M> ecal_subscriber_;
    ..... 
    ecal_subscriber_.Create(std::string(topic_name));
    ecal_subscriber_.AddReceiveCallback(std::bind(&GenericSubscriber::eCALCallback, this, 
    std::placeholders::_1, std::placeholders::_2));
    auto evt_callback = std::bind(&GenericSubscriber::OnEvent, this,std::placeholders::_1, 
    std::placeholders::_2);
    ecal_subscriber_.AddEventCallback(sub_event_connected, evt_callback);
    ecal_subscriber_.AddEventCallback(sub_event_disconnected, evt_callback);
    ecal_subscriber_.AddEventCallback(sub_event_dropped, evt_callback);
    ecal_subscriber_.AddEventCallback(sub_event_timeout, evt_callback);
    ecal_subscriber_.AddEventCallback(sub_event_corrupted, evt_callback);
```
the OnEvent callback implemented as bellow (just paste code from ecal sample):
```
void GenericSubscriber<T,M>::OnEvent(const char* topic_name_, const struct eCAL::SSubEventCallbackData* data_) {
  std::cout << "topic name   : " << topic_name_ << std::endl;
  switch (data_->type) {
    case sub_event_connected:
      std::cout << "event        : "
                << "sub_event_connected" << std::endl;
      break;
    case sub_event_disconnected:
      std::cout << "event        : "
                << "sub_event_disconnected" << std::endl;
      break;
    case sub_event_dropped:
      std::cout << "event        : "
                << "sub_event_dropped (" << data_->clock << " messages)" << std::endl;
      break;
    case sub_event_timeout:
      std::cout << "event        : "
                << "sub_event_timeout" << std::endl;
      break;
    // not implemented yet
    case sub_event_corrupted:
      std::cout << "event        : "
                << "sub_event_corrupted" << std::endl;
      break;
    default:
      std::cout << "event        : "
                << "unknown" << std::endl;
      break;
  }
  std::cout << std::endl;
}
```

**To Reproduce**
compiled done and run my node, I can see the log output as below and everything are ok:
```
topic name   : /generic_tf/odom_center_of_rotation
event        : sub_event_connected

topic name   : /odom
event        : sub_event_connected

topic name   : /imu/data_raw
event        : sub_event_connected
topic name   : /scam
event        : sub_event_connected
```

after a while,I  got the errors as below:

```
topic name   : /imu/data_raw
event        : sub_event_dropped (21 messages)

event        : sub_event_dropped (20 messages)
topic name   : /scam
event        : sub_event_dropped (15 messages)

topic name   : /odom
event        : sub_event_dropped (20 messages)

topic name   : /generic_tf/odom_center_of_rotation
event        : sub_event_dropped (143 messages)

```
As a comparison
I developed another node,just simply subscribe the same topics  /scan (the frequency is 10hz),/odom (the frequency is 100 hz),/imu/data_raw(the frequency is 100 hz),/generic_tf/center_of_rotation_2_odom (the frequency is 100hz). the program runs well
and do not output the event "sub_event_dropped".

I have no idea what can cause the messages to drop in the cartographer_ros node.

**Develop enviroment**
 - OS: [Ubuntu 16.04]
 - ecal global configuration(/etc/ecal/ecal.ini): default

**Comment by rex-schilasky** [comment]
Hi @long131ss, the local shared memory communication is loosely coupled by default. So the publisher is not delayed (synchronized) by the processing of the message by the receiving subscriber. That means .. if your subscriber callback is blocked (processing the last received message) it can not receive a new incoming message and the internal drop counter will increase.
So if your subscribing process is slowing down (resource leaks, queueing, internal threading ..) such a delayed dropping could occur.
You can force eCAL to sync a publish/subscribe connection (publisher is blocked by a slow subscription) by setting a timeout parameter in the global ecal.ini file (see #441).

**Comment by long131ss** [comment]
hi @rex-schilasky 

thanks for the reply.

sorry for misleading, but what I'm experiencing is not the 'sync problem between pub & sub', I'll try to describe the issue again below:

I have different nodes doing publishing constantly, and I wrote another node to subscribe to these subscribers(let's say 2), to do the test, I created an infinite loop in one of the subscriber's callback functions, what I expected is 'The other subscriber's callback could still be called in this situation, but it's NOT.
```
// define a subscriber callback function
void OnReceive1(const char* topic_name_, const struct eCAL::SReceiveCallbackData* data_, const void* par_)
{
  while(true) 
  {
    printf("we sleep in OnReceive1");
    sleep(1);
  }
  
}
void OnReceive2(const char* topic_name_, const struct eCAL::SReceiveCallbackData* data_, const void* par_)
{
  printf("We received %s on topic %s\n.", (char*)data_->buf, topic_name_.c_str());
}
// create 2 subscriber
eCAL::CSubscriber sub1("foo1", "std::string");
eCAL::CSubscriber sub2("foo2", "std::string");
// register subscriber callback function
sub1.AddReceiveCallback(OnReceive1);
sub2.AddReceiveCallback(OnReceive2);
 ```
In the above example,`OnReceive1` is printing sleeping while `OnReceive2` would not be called at all.

It seems like if I block one subscriber, all the other subscribers would be blocked as well, is this true?

if this is the case, is there any way for me to get the subscriber callbacks invoked asynchronously?

Thanks and looking forward to your feedback.

**Comment by long131ss** [comment]
okay, so after digging into the source code, I found the subscribers' callbacks are actually involved in different threads, BUT while involving the callbacks, they are sharing the same mutex in 'ecal/core/src/pubsub/ecal_subgate.cpp' : 
```
  size_t CSubGate::ApplySample(const std::string& topic_name_, const std::string& topic_id_, const char* buf_, size_t len_, long long id_, long long clock_, long long time_, size_t hash_, eCAL::pb::eTLayerType layer_)
  {
    if(!m_created) return 0;

    // update globals
    g_process_rclock++;
    g_process_rbytes_sum += len_;

    // add sample to data reader
    size_t sent(0);
    std::lock_guard<std::mutex> lock(m_topic_name_datareader_sync);
    auto res = m_topic_name_datareader_map.equal_range(topic_name_);
    for (auto it = res.first; it != res.second; ++it)
    {
      sent = it->second->AddSample(topic_id_, buf_, len_, id_, clock_, time_, hash_, layer_);
    }

    return sent;
  }
```

Could you elaborate a bit more regarding the lock here? `m_topic_name_datareader_sync`, thank you!

**Comment by rex-schilasky** [comment]
Hi, 

yes this mutex protects the data reader map against data races. The `AddSample` function is the central point of message injection that is called from the different transport layers (shm, udp, innerproc) and does a synchronous call of the topic attached callbacks. So if you block on callback the other ones will not be triggered anymore.

The threading model is described [here](https://continental.github.io/ecal/advanced/threading_model.html). So even there are multiple internal receive thread handling udp and shm connections finally all is synched here ...

**Comment by andersdu** [comment]
Hi,

I happen to have this same requirement.

Do you mean if a topic has different pub/layer sources, this mutex is designed to protect this case?

Do you have any suggestions for get around this?

I need this because there's indeed cases the subscribers' callback would be blocked, i.e. the data processing part would lock the same mutex of callback to avoid data races.

Thanks.

**Comment by rex-schilasky** [comment]
Hi @long131ss, Hi @andersdu,

we discuss that now internally. There are different options we could implement like a decoupled thread calling the users callback with a copy of that data or maybe comparable solutions. Let's see what can be done without decreasing performance and doing extra copies or queuing data.

For now you can for sure make a copy of the received message buffer and process that copy in your own thread. So the callback is only blocked for the time needed to make that copy and you can process asynchronously - at least as an intermediate solution that could fit. What do you mean ?

Best, ReX

**Comment by andersdu** [comment]
thanks @rex-schilasky 

Basically the current version of my code is doing what you suggested: "make a copy of the received message and process the copy in another thread", the problem is this "processing" could sometimes take longer time, which caused the "make a copy of the received message" blocked(you don't want to change the copy while it's being used).

currently I'm working with removing the `m_topic_name_datareader_sync`, and it seems works fine maybe because I don't have different layers of publishers to same topic.

Maybe you could implement the "lock" to only protect for same topic, instead of blocking other topics' callbacks?

Br.
Anders

**Comment by rex-schilasky** [comment]
Hi,

issue is fixed now by using a shared lock. The fix will be part of the 5.10 version. 

Thanks for reporting, ReX
